pkg/apis/query/v0alpha1/query.go

```diff
@@ -0,0 +1,212 @@
+	// It can be used to distinguish different types of queries.
+	QueryType string `json:"queryType,omitempty"`
+
+	// MaxDataPoints is the maximum number of data points that should be returned from a time series query.
+	MaxDataPoints int64 `json:"maxDataPoints,omitempty"`
+
+	// Interval is the suggested duration between time points in a time series query.
+	IntervalMS float64 `json:"intervalMs,omitempty"`
+
+	// true if query is disabled (ie should not be returned to the dashboard)
+	// Note this does not always imply that the query should not be executed since
+	// the results from a hidden query may be used as the input to other queries (SSE etc)
+	Hide bool `json:"hide,omitempty"`
+
+	// Additional Properties (that live at the root)
+	props map[string]any `json:"-"`
+}
+
+// TimeRange represents a time range for a query and is a property of DataQuery.
+type TimeRange struct {
+	// From is the start time of the query.
+	From string `json:"from"`
+
+	// To is the end time of the query.
+	To string `json:"to"`
+}
+
+func (g *GenericDataQuery) AdditionalProperties() map[string]any {
+	if g.props == nil {
+		g.props = make(map[string]any)
+	}
+	return g.props
+}
+
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
+func (g *GenericDataQuery) DeepCopyInto(out *GenericDataQuery) {
+	*out = *g
+	if g.props != nil {
+		out.props = runtime.DeepCopyJSON(g.props)
+	}
+}
+
+// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new GenericDataQuery.
+func (g *GenericDataQuery) DeepCopy() *GenericDataQuery {
+	if g == nil {
+		return nil
+	}
+	out := new(GenericDataQuery)
+	g.DeepCopyInto(out)
+	return out
+}
+
+// MarshalJSON ensures that the unstructured object produces proper
+// JSON when passed to Go's standard JSON library.
+func (g GenericDataQuery) MarshalJSON() ([]byte, error) {
+	vals := map[string]any{}
+	if g.props != nil {
+		for k, v := range g.props {
+			vals[k] = v
+		}
+	}
+
+	vals["refId"] = g.RefID
+	if g.Datasource.Type != "" || g.Datasource.UID != "" {
+		vals["datasource"] = g.Datasource
+	}
+	if g.DatasourceId > 0 {
+		vals["datasourceId"] = g.DatasourceId
+	}
+	vals["intervalMs"] = g.IntervalMS
+	vals["maxDataPoints"] = g.MaxDataPoints
+	return json.Marshal(vals)
+}
+
+// UnmarshalJSON ensures that the unstructured object properly decodes
+// JSON when passed to Go's standard JSON library.
+func (g *GenericDataQuery) UnmarshalJSON(b []byte) error {
+	vals := map[string]any{}
+	err := json.Unmarshal(b, &vals)
+	if err != nil {
+		return err
+	}
+	key := "refId"
+	v, ok := vals[key]
+	if ok {
+		g.RefID, ok = v.(string)
+		if !ok {
+			return fmt.Errorf("expected string refid (got: %t)", v)
+		}
+		delete(vals, key)
+	}
+
+	key = "datasource"
+	v, ok = vals[key]
+	if ok {
+		wrap, ok := v.(map[string]any)
+		if !ok {
+			return fmt.Errorf("expected datsource as object (got: %t)", v)
```

Comment:
Nit. Legacy queries might have `"datasource": "<datasource name>"`?

Reply:
good point -- i'll add some more examples and tests to the parser